### ISSUE TITLE
Weird problem with 0.7.3

### ISSUE THREAD

**Issue by ShaheedHaque** [jpype-project/jpype#720]
We've been happily using JPype on Ubuntu 18.04 for a while (for which many thanks!), across variety of versions including most recently 0.7.1 and 0.7.2. We recently tried to deploy our software with 0.7.3 and found some confusing test failures with few external signs of what might be wrong:

- At least one instance of a process dying with a SIGABRT.
- We have a script we use to start and stop our code via systemd. The start/stop commands normally complete in 15-20 seconds. After a failure, the "stop" takes MUCH longer to run - perhaps one minute.
- Nothing obvious in "top" or /var/log/syslog or our system-specific logs.

We realise the above points are not very specific, but we were able to confirm by flipping between 0.7.2 and 0.7.3 that the latter is the problem.

Apologies for the rather scant report, please advise if there is some debugging steps we can assist with.

Thanks, Shaheed

P.S. We can pin to 0.7.2 as a workaround. We note that 0.7.4 has been released but we have not tried  it as it has not made its way to PyPi yet. Also, it seems to features some memory-related fixes for arrays, but we are not using arrays.

**Comment by Thrameos** [comment]
Well you will use arrays whether you intend to or not.  All Java variable argument function implicitly create an array whenever they are called.   Which is unfortunately, why a hot fix was required.  

As for your issue, I would need some additional diagnostics.    We have been working on shutdown related issues, but neither of these releases have a shutdown enabled.   Thus, I am going to operates on the assumption that this is related to the class hints changes made between 0.7.2 and 0.7.3.   It would be great if you could get me some additional diagnostics.    I have one other reported crash in 0.7 series but that is tied directly to shutdown issues.  

As for the shutdown issue.  I recommend using something like jstack to find out what threads are running.  If you can get a backtrace off a core dump it would be most helpful.

**Comment by ShaheedHaque** [comment]
On the arrays, understood.

Sadly we have no coredumps. Is there some backdoor way to trigger one? Or perhaps some way to dump the heap of a running process?

**Comment by Thrameos** [comment]
Internally Java does use threads.  Internally JPype has 3 threads of interest.  The main thread which the user sees.  The reference queue which does clean up of Java references.  Last the Java finalize queue is also active.  

As far as producing core dumps, the only thing I can recommend is perhaps ulimit.  The heap on the running process does little good as the problem is unlikely to be visible.  Perhaps you can trigger the problem by calling "shutdownJVM".   Or run the process under a debugger.  The pattern for a debug run is in the azure_pipelines.     

For example to find a hard to trace segfault on a remote machine I used this pattern.
```
gdb -ex 'handle SIGSEGV nostop noprint pass' -ex "run -m pytest -v test/jpypetest/test_jstring.py --checkjni"  -ex "bt" -ex "quit" python
```

This calls ``python -m pytest -v test/jpypetest/test_jstring.py --checkjni`` with a debugger watching for problems.

**Comment by ShaheedHaque** [comment]
Unfortunately, the gdb thing is not really a runner as the suspect process is a "worker" started by Celery main process, and there is no control over how the worker is spawned. (Hacking Celery's multiprocessing layer is not for the faint hearted, and is something I'd prefer to avoid).

I'll ponder the shutdownJVM. We don't use it at present, but Celery is set up to restart its worker processes after each has done X amount of work.

Back on arrays, it seems I had misremembered: we *do* in fact use arrays (almost exclusively...):

- the data we are passing into JPype is a non-trivial Python object (as stringified JSON, we start around 50 to 100kB, and can reach 10s of MB).
- we get back a PDF as a byte array

```
        ds = JsonQLDataSource(io.ByteArrayInputStream(jpype.JArray(jpype.JByte)(bytearray(data, 'utf-8'))), query)
        jasper_print = JasperFillManager.fillReport(report, parameters, ds)
        JasperExportManager.exportReportToPdfStream(jasper_print, tmp_stdout)
        #
        # Emit PDF.
        #
        return bytearray(b & 0xff for b in tmp_stdout.toByteArray())
```

So, perhaps the fix in 0..7.4 is what we need?

**Comment by Thrameos** [comment]
Yeah.  Lets start there.  Assuming that you worker processes were hitting the resource limit it would cause odd crashes and long cleanup periods.  That means we should test with 0.7.4.

**Comment by ShaheedHaque** [comment]
I grabbed 0.7.4 and am delighted to say that it seems to work as expected/hoped. (I'll do a couple more runs just to be sure, but I did not want you to waste time on this).

**Comment by ShaheedHaque** [comment]
I spoke too soon. I'm getting some results which are so anomalous that I cannot quite be sure what is going on. I'll do some more on this later and report back.

**Comment by Thrameos** [comment]
Any update on this?

**Comment by ShaheedHaque** [comment]
Sorry, it is taking a while to put some other fires out.

**Comment by Thrameos** [comment]
Something appears to have broken in our release cycle process.  Two different commit points are being distributed as JPype 0.7.4.   One of them has shutdown logic which was under development.  Not sure how it happened exactly but it may explain some issues.

**Comment by marscher** [comment]
I have uploaded the binary wheels from github to pypi. The source archive is a checkout of the tagged version. It is absolutely unclear to me how this could have happened.
I suggest, that we switch to use a modern git based versioning for setup.py (which contains the last tag, the git hash and so on). The current toolset makes it hard or impossible to figure out, which version (commit) is actually used.

@ShaheedHaque did you install from source or binary wheel?

**Comment by ShaheedHaque** [comment]
@marscher From pypi.org. I didn't watch closely if it was built on the fly or just pure download (I have a full jdk setup so it could have built without any problems)

**Comment by marscher** [comment]
when you are using pip to install without any further arguments, it will prefer the binaries, which are provided for your platform.

**Comment by marscher** [comment]
@Thrameos the usage of binary wheels makes me think, that the binaries have been built from master instead of the 0.7.4 tag.

**Comment by Thrameos** [comment]
I checked the source distribution on PyPi and it is from master.   I then pulled both of the source distributions from the release... Neither was from master.  I then pulled each of the wheel files from the wheelhouse and grepped for "JTerm" in jpype.   None have it (they are from release).   The size of the source distribution on PyPi doesn't match any of the versions that in the docker build area, but does match the one on the release site.

The checksums for all the binaries match but the source file does not.  So the likely source of the issue is a bad source distribution.

**Comment by Thrameos** [comment]
I do not see the offending line in the version difference tool.  So a version pulled from the tag should not have produced a bad file.

https://github.com/jpype-project/jpype/compare/v0.7.3...v0.7.4

My best guess is somewhere the source file got corrupted and that then got pushed to conda which generated a series of bad binaries.  So the question is how do we correct this?

I can tag v0.7.4 again as v0.7.5 and rebuild all the distributions with docker.   But that will take some time before I will be free again.

**Comment by Thrameos** [comment]
I completed verification, the bad distribution appears to be the source version on PyPi and Conda all versions.

**Comment by marscher** [comment]
The conda build is pulled of the pypi source. So we have to fix the source, and then trigger a new build for conda.
On pypi there is a policy saying that you cannot change an existing file, but we can upload a different archive format. We have to delete the tar.gz and replace it with a zip.
I'm really sorry for all the trouble this has caused.

**Comment by Thrameos** [comment]
Okay,   So my plan is to start testing out the azure release pipeline this evening.   

- [ ] I will create a releases/0.7.5 branch off of the 0.7.4 tag.  
- [ ] I will then set the build pipeline to publish a Source distribution as an artifact (only off of a releases/# push so we only get artifacts off of an intended release.)
- [ ] I will have it build each of the major platforms.  (probably  about 20 commits to get the pipeline fully running)
- [ ] Once that is complete I will squash all the commits to a single commit (it will be a mess, I guarantee it... try.. try... what gives.... you can't be serious.... really... try again...typo... this time for sure...)
- [ ] I will update the release numbers accordingly so that we put out a clean release.
- [ ] I will make a new pull request for your approval from the release branch to master so that you can verify the updates and approve the release.
- [ ] Once I have the approval, I will tag a v0.7.5 release on the release page.  (with the MD5SUMs from the proper release) 
- [ ] You can the start the pushing process.
- [ ] We can work together to get the release pipeline have  publish to pypi once we believe the rest of the process is sound (on a manual trigger so it will only publish when approved).

Don't worry too much about it.... We were always going to have to get the release system working.  It was just a matter of time before this time bomb when off.  Better to mess up a maintenance release than the big 0.8.0.

Does that sound like a good plan?

**Comment by Thrameos** [comment]
Wow took me the entire night just to update the azure scripts (didn't even get to test them).   Hopefully working tomorrow.

**Comment by marscher** [comment]
Sounds good to me. Especially automatic deployment. It could be a risk storing encrypted credentials on azure. On Travis there was a flaw  once and credentials got leaked.
Eventually we should also consider to sign the files with gpg.

On May 10, 2020 1:31:46 AM UTC, Karl Nelson <notifications@github.com> wrote:
>Okay,   So my plan is to start testing out the azure release pipeline
>this evening.   
>
>- [ ] I will create a releases/0.7.5 branch off of the 0.7.4 tag.  
>- [ ] I will then set the build pipeline to publish a Source
>distribution as an artifact (only off of a releases/# push so we only
>get artifacts off of an intended release.)
>- [ ] I will have it build each of the major platforms.  (probably 
>about 20 commits to get the pipeline fully running)
>- [ ] Once that is complete I will squash all the commits to a single
>commit (it will be a mess, I guarantee it... try.. try... what
>gives.... you can't be serious.... really... try again...typo... this
>time for sure...)
>- [ ] I will update the release numbers accordingly so that we put out
>a clean release.
>- [ ] I will make a new pull request for your approval from the release
>branch to master so that you can verify the updates and approve the
>release.
>- [ ] Once I have the approval, I will tag a v0.7.5 release on the
>release page.  (with the MD5SUMs from the proper release) 
>- [ ] You can the start the pushing process.
>- [ ] We can work together to get the release pipeline have  publish to
>pypi once we believe the rest of the process is sound (on a manual
>trigger so it will only publish when approved).
>
>Don't worry too much about it.... We were always going to have to get
>the release system working.  It was just a matter of time before this
>time bomb when off.  Better to mess up a maintenance release than the
>big 0.8.0.
>
>Does that sound like a good plan?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jpype-project/jpype/issues/720#issuecomment-626258803

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by Thrameos** [comment]
I assume you mean "except automatic deployment" .   We don't have to have azure actually do a push.  It can just push the candidates someplace that we can verify and sign so azure won't actually have the real credentials.

Still a few hours from complete on this task.  Windows appears to have some issue with setting the JVM so it works if I don't set to a particular version, but not a great option in the long run as I want everything in the build system pinned.  OSX builds, so I just have to get the windows resolved and finish the Linux.  

I wrote up a release guide in the doc directory so that the procedure it defined.  Should be wrapped up to today so that I can work on the remaining 0.8 items during the week.

**Comment by ShaheedHaque** [comment]
If someone could kindly confirm when the results of all this effort hits pypi.org, I'll look to do some testing on my end.

**Comment by Thrameos** [comment]
The release copies are now published on github.  I assume that Martin will do the pushes to PyPi and trigger a conda cycle sometime tomorrow.  I am sorry for any problems that this bad release caused.  I am still working toward a 0.8 release sometime in the next month.   If you could be so kind as checking it out once we have a beta release ready to go.

**Comment by marscher** [comment]
I have uploaded the new source archive to pypi yesterday.
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by ShaheedHaque** [comment]
Cool. I'll try to get some testing done tomorrow...

**Comment by Thrameos** [comment]
I know the source distribution is updated, but 0.7.5 has not yet hit PyPi or conda.  Please verify the version you are testing does not have the string “Term” in jpype/_core.py.

**Comment by ShaheedHaque** [comment]
I deployed with 0.7.5 (and checked there is no "Term" in _core.py), and still hit this mysterious issue. As before, the Celery process seems "stuck", taking ages to kill.

(And elsewhere, 0.7.2 seems to continue working without issue).

**Comment by ShaheedHaque** [comment]
Idle thought. Celery does strange and magical things with regard to multiprocessing (it does not use the standard Python multiprocessing library, but rather its own extension called ["billiard"](https://github.com/celery/billiard)). Does that ring any bells?

**Comment by Thrameos** [comment]
Unfortunately that one is new to me.   Is there a way you can create a skeleton program that replicates the hang.   I can fix it pretty quick if you can pull that off.

I have a work around that will allow you to force exit the JVM but it will leave Python in a less than desirable state which is fine if the worker is just going to exit.  But lets try to fix it first.

**Comment by ShaheedHaque** [comment]
Celery takes a bit of driving; I need to think how to do this.

**Comment by Thrameos** [comment]
I think I have a theory.   The way the JVM is structure it must always be started from the main thread.  When it is shutdown, all other attached threads are pinged to shutdown.  Once the last thread which is not main is gone it proceeds with the shutdown sequence.  If for some reason your code has started the JVM from another thread, or has attached other threads as non-daemon then it would hang.

There are a few possible work arounds for this if this is actually the problem.
1) Start JPype from the main thread if possible.
2) If not possible make sure that you attach all other threads as daemon (Do not use attachThreadToJVM or you risk triggering this problem.)
3) If you want to shut a process down without triggering a clean shutdown use Runtime.halt().

Please note #733 is required.   Currently JPype is attaching all threads automatically as non-daemon which is likely to trigger this problem.   If you don't have time now, we can check again one integration testing is started.  Hopefully once all the patches are pulled together we can solve this.

**Comment by ShaheedHaque** [comment]
On Thu, 28 May 2020 at 05:39, Karl Nelson <notifications@github.com> wrote:

> I think I have a theory. The way the JVM is structure it must always be
> started from the main thread. When it is shutdown, all other attached
> threads are pinged to shutdown. Once the last thread which is not main is
> gone it proceeds with the shutdown sequence. If for some reason your code
> has started the JVM from another thread, or has attached other threads as
> non-daemon then it would hang.
>
> There are a few possible work arounds for this if this is actually the
> problem.
>
>    1. Start JPype from the main thread if possible.
>    2. If not possible make sure that you attach all other threads as
>    daemon (Do not use attachThreadToJVM or you risk triggering this problem.)
>    3. If you want to shut a process down without triggering a clean
>    shutdown use Runtime.halt().
>
> Please note #733 <https://github.com/jpype-project/jpype/pull/733> is
> required. Currently JPype is attaching all threads automatically as
> non-daemon which is likely to trigger this problem. If you don't have time
> now, we can check again one integration testing is started. Hopefully once
> all the patches are pulled together we can solve this.
>

I can see how that could explain the long shutdown. In my case, I'm not
explicitly using Python threads or OS threads; specifically:

- The Celery folk don't use threads (I've asked this question in another
context, and the answer was clear).
- I start JPype from a declarative context, i.e. at module scope:

import jpype
MAXMEM = getattr(settings, 'PAIYROLL_DEBUG_JVM_MAXMEM', '512M')
TMPDIR = tempfile.mkdtemp(prefix='jvm-' + temp.PREFIX)
jpype.startJVM('-Xmx' + MAXMEM, '-Djava.awt.headless=true',
'-Djava.io.tmpdir=' + TMPDIR, classpath=CLASSPATH,
               convertStrings=False)
from jpype import java

This must, I presume be on both Python's main thread and the main OS
thread. As you can probably guess, I don't explicitly shut down the JVM
either. I *do* have an atexit() handler, but its purpose is to tidy up some
of the temporary crap that the JVM is prone to leave lying around:

@atexit.register
def atexit_handler():
    """
    TMPDIR pruning.

    @atexit.register calls us on Ctrl-C or SIGINT.
    """
    try:
        shutil.rmtree(TMPDIR)  <<<< note this TMPDIR is the sameas was
passed into startJVM().
        logger.debug(_('JVM cleaned up {0} atexit').format(TMPDIR))
    except FileNotFoundError:
        logger.debug(_('JVM already cleaned up {0} atexit').format(TMPDIR))

So, could the lack of a call to Runtime.halt() (or rather perhaps a
putative Python proxy) cause the shutdown issue?

Either way, it is definitely easier for me to test a new version than build
a cut down test case (the latter is still on my TODO list).

Thanks, Shaheed

**Comment by Thrameos** [comment]
Okay this gives me some more context to work with.   An at exit could definitely play a role.  So I will need to consider it.  

Does celery use processes?  If so do they spawn or fork them?  Forking is a big problem as the JVM does not properly survive a fork.  But it that were the case you would have other noticeable things happening.   I will ponder it more after work.

Lets hold off cutting down a version until I have an integration test of JPype ready.   I have a huge pile of bug fixes that I am pulling together and any one of them may actually address the problem.  So if it is easiest for you I will work to pull together a release candidate (with an option for diagnostics to help you out).

I believe the diagnostics required will be a dump of open JDK threads, a dump of shutdown hooks, a shutdown logger,  a time check for each stage of the shutdown sequence.  Then you can trigger diagnostics from you code send the log.  That will get us a handle on the problem.

**Comment by ShaheedHaque** [comment]
On Thu, 28 May 2020, 17:07 Karl Nelson, <notifications@github.com> wrote:

> Okay this gives me some more context to work with. An at exit could
> definitely play a role. So I will need to consider it.
>
> Does celery use processes? If so do they spawn or fork them? Forking is a
> big problem as the JVM does not properly survive a fork. But it that were
> the case you would have other noticeable things happening. I will ponder it
> more after work.
>

I'm not 100% sure  as the code is not simple (
https://github.com/celery/billiard/tree/master/billiard) but I think it
works like this: fIrst, Celery has a main process which, by default, uses
fork (
https://stackoverflow.com/questions/24267127/celery-task-subprocesses-fill-up-concurrency-slots)
to create N workers. Each worker then runs "my code" by looking up the
Python entry point and calling into it. I guess if this is a problem, then
the fact that it worked up through 0.7.2 was a fluke?


> Lets hold off cutting down a version until I have an integration test of
> JPype ready. I have a huge pile of bug fixes that I am pulling together and
> any one of them may actually address the problem. So if it is easiest for
> you I will work to pull together a release candidate (with an option for
> diagnostics to help you out).
>
Ack. No quibble with that approach.


> I believe the diagnostics required will be a dump of open JDK threads, a
> dump of shutdown hooks, a shutdown logger, a time check for each stage of
> the shutdown sequence. Then you can trigger diagnostics from you code send
> the log. That will get us a handle on the problem.
>
I'm in your hands.

Thanks, Shaheed

**Comment by Thrameos** [comment]
If a process is being created with fork which has a live JVM then it is living dangerously.   I never got that to work because the JVM always seemed to have some resource in a bad state.   That is not to say we can't make it create a bunch of processes from with JVMs.   The testbench in JPype does that spinning up a bunch Python copies each with spawn their own JVM for testing.  

It is not hard to get Python multiprocess to generate a spawn copy rather than a fork copy but not sure how nice that will be with celery.  I doubt that Celery is actually using any resource from before the fork as most times you just take the ends of the pipe and call exec.   So the differences between spawn (launch an executable) and fork (create a full process copy) is likely small.

I suppose as a safety feature I could record the process id of the original process so that if you attempt to run a fork it will detect it an act accordingly (defensively).  That would make sure the user is always directed towards a usable option.  I am always for inherent safety in toolkit if it doesn't like completely slaughter performance.  Checking on a few critical calls (like say creating a new class) would get 95% of the problem cases.

**Comment by ShaheedHaque** [comment]
On Thu, 28 May 2020 at 18:58, Karl Nelson <notifications@github.com> wrote:

> If a process is being created with fork which has a live JVM then it is
> living dangerously. I never got that to work because the JVM always seemed
> to have some resource in a bad state. That is not to say we can't make it
> create a bunch of processes from with JVMs. The testbench in JPype does
> that spinning up a bunch Python copies each with spawn their own JVM for
> testing.
>
Equally, I think it would not be hard for me to defer initialization of
JPype until I knew I had Java work to do. (nce Celery does not use threads,
I'm reasonably sure that could be guaranteed to be on the main OS thread).

**Comment by ShaheedHaque** [comment]
I'm not yet sure if this is significant, but here are the results of some A-B tests of simply starting and stopping our application stack (multiple Python programs managed via systemd) firstly with 0.7.2:

```
$ time ./bin/stop 

real    0m14.102s
user    0m2.703s
sys     0m0.323s
$ time ./bin/start

real    0m7.255s
user    0m2.764s
sys     0m0.349s
$ time ./bin/stop 
... the above repeated 4 times gives an average start time of about 7 seconds, and a stop time of 14s.
```

Now replace 0.7.2 with the putative 1.0.0:

```
$ time ./bin/start

real    0m7.860s
user    0m3.889s
sys     0m0.377s
$ time ./bin/stop

real    0m36.548s        <<<<<< 36s instead of about 14s
user    0m3.998s
sys     0m0.372s
$ time ./bin/start

real    0m13.230s        <<<<< 13s instead of about 7s
user    0m4.153s
sys     0m0.377s

$ time ./bin/stop

real    0m14.491s
user    0m3.618s
sys     0m0.389s
$ time ./bin/start

real    0m8.466s
user    0m3.970s
sys     0m0.434s
$ time ./bin/stop

real    0m16.500s
user    0m4.040s
sys     0m0.312s
... and now it has returned more or less back to the original 7 + 14 pattern (OK, maybe an extra second added to each).
```

In other words, after an initial/one-off increase in stop times, the system subsequently starts and stops roughly as expected. Now it is worth repeating that there was no "activity" involving JPype (or indeed any other part of our application) other than starting or stopping.

Since the increase in stop time was previously thought to be the one definitive marker of unwanted behaviour, I will now try running actual work through the system to see if there is anything more substantive amiss.

**Comment by marscher** [comment]
The increased stop times are due to the newly introduced shutdown sequence, which cleans up acquired resources. So I guess this is totally expected from now on. Without any sane statistics it would be hard to make the point, that the startup procedure is taking longer than in older versions.

**Comment by Thrameos** [comment]
The old version does not actually shut anything down.  It just halts the JVM without destroying resources other than removing Python references.  Between that and the conversion to use CPython classes exclusively which was an increase in start up costs, the numbers don't surprise me that we are seeing slower starts and shutdowns.  We definitely have suffered a significant start up penalty now that we load all the required classes at once at the start and we are having much more activity in terms of clean up.   But I am not sure that 36s would be reasonable.  But that isn't a deadlock so it means something must be holding up the shutdown sequence without actually causing it to fail.

The tradeoff with the startup and shutdown costs being that the run speed of 0.7.5 is much faster in terms of JNI calls than most previuos versions.  If you can provide a short piece of sample code that replicates it I may be able to identify the source of the issue.  I have done abut 5 passes through the shutdown logic and am reasonably sure there is no fatal deadlocks.  But there are race conditions in which the Java sequence has to wait until a Proxy is closed and that means depending on Python timeouts (as opposed to dropping the Proxy while running and hope that it doesn't crash on the way out later.)

You are still able to call JVM halt function which should terminate the process without any further processing.   It will terminate the Python process as well.  You must make sure all files are flushed and closed prior to calling the halt method.

[Halt vs Exit](https://www.baeldung.com/java-runtime-halt-vs-system-exit)

**Comment by ShaheedHaque** [comment]
To be clear, I'm not complaining about the start and stop times... It is not a problem for us. What I found interesting is that the new code only exhibited the slow start and stop once. I guess os-level caching must be at play.

Anyway, I ran our test suite repeatedly against the new code (12 times in fact) and saw no failures of the kind originally reported. In comparison, 0.7.4 failed even a single pass, and 0.7.5 failed on the 3rd or 4th pass. I have to think that means the issue of locking up Celery is fixed. 

At any rate I'll switch my everyday dev setup to the new code for some extended tests too. (I am being ultra cautious as I'm aware I declared victory prematurely before).

**Comment by ShaheedHaque** [comment]
It seems we are not yet out of the woods on this with 1.0.1:

- As reported above, I did get 12 back-to-back ok-seeming runs. Each "run" here consisted of booting a fresh VM, and checking firstly that our stack came up, and secondly that it responded to a "stop" in a reasonable amount of time (i.e. a simple start/stop test).
- With continued ad-hoc testing on our local dev machines, we have one Linux scenario that seems related to this issue and my Mac-based colleague has a second where he just started seeing some JVM-based crashes. We are trying to get a solid repro for the former, and a full core dump for the latter.

**Comment by Thrameos** [comment]
I have another fix regarding the shutdown.  I tried to push the resource destruction later in the cycle in #832.   Unfortunately it is difficult for me to be much further assistance without some diagnostics of where the crash is occurring.  There are a lot of possibilities and I will need it narrowed down a bit.

**Comment by ShaheedHaque** [comment]
> I will need it narrowed down a bit.

Understood.

After some debugging with JPype 1.0.2, I have one solitary fact so far: the Celery workers which seem to be "stuck" are not responsive to a SIGTERM but can be killed by SIGKILL. On one occasion, 22 were killed by the SIGTERM, and when the remaining 2 were SIGKILLed, the system as a whole resumed operation (the death of the workers is noticed by the main process, which starts new replacement workers).

**Comment by ShaheedHaque** [comment]
I tried to use the faulthandler module to get some stack traces for Python threads, and noticed that only the Celery main process ends up generating the output file I specify; I am pretty sure this confirms that the workers are forked, and not spawned.

**Comment by ShaheedHaque** [comment]
OK, some solid progress. I moved the JPype initialisation from module scope to delay it until actually needed, and this allowed - for the first time since 0.7.2 - a full run of our test suite. None of my earlier reports of partial success have come anywhere near this, so it represents real progress.

It also speaks to @Thrameos assertion that forking might be problematic :-). More testing will follow, but "so far, so good".

Curiously, the little Python program I use to wrap our invocation of pytest exited like this (having run the full test suite successfully):

```
Fatal Python error: Segmentation fault

Thread 0x00007f84e4194740 (most recent call first):
  File "/usr/local/lib/python3.8/dist-packages/jpype/_core.py", line 321 in _JTerminate
```
And here is line 321:
```
# In order to shutdown cleanly we need the reference queue stopped
# otherwise, we can experience a crash if a Java thread is waiting
# for the GIL.
def _JTerminate():
    try:
        _jpype.shutdown()  <<<<<<<<<<<<< 321
    except RuntimeError:
        pass
```
The *only* reference to jpype that this program could have made is an import as part of its transitive fanout:
```
import jpype
```
and it won't have invoked startJVM(). This is at worst a minor annoyance, but thought you might like to look into this.

**Comment by ShaheedHaque** [comment]
OK, moving startJVM to after the fork of the worker has done the trick and gets us working on 1.0.2, subject to the SEGV on shutdown. @Thrameos: Do you want me to close this issue and raise another to track that?

**Comment by Thrameos** [comment]
Yes, lets start a fresh one.   My development machine is down so I can't be much help for a week or two until I do much.
